### ISSUE TITLE
Refactor load and storage functionality to be more generic and robust

### ISSUE THREAD

**Issue by franzlst** [DLR-RM/RAFCON#369]
So after already struggling a while with our storage and file system path and especially the get_file_system_path method of the State-Class I will propose a re-factoring by this issue to achieve the following properties:
- load of states (core data and meta data) without any knowledge of folder structure (only the file name has to be of the right kind and naming e.g. core_data.json, meta_data.json and state_machine.json.
- storage_path methods are not called to get the file_system_path (what was a very bad idea in the past
- no set_file_system_path anymore because the file system path has to be set by the storage method load_* or save_* 
- models can identify by there-self if there is maybe meta data of a previous stored state (core + gui)
- very simple pattern which is easily to understand
- no more minor incompatibility problems because of the folder naming structure

Small extra point that is related:
- the reduction/removal of mark_for_removal list of state paths in the storage module

Therefore get_file_system_path will become a property that only can be read by the public interface and set via direct access to the semi-private \_file_system_path variable or method. This variable is only set by the storage module and its save_* or load_* functions. The get_file_system_path method hereby will be reduced to a clean property without any if, elif and else condition checks and so on.

If the file system path is set and created without concatenating before loading or storage this strings is also independent from state folder naming issues. The naming of state folders can be even mixed and be cleaned up by the next storage command. The type of folder naming can be even left to the user by an option or kind of configuration.

The mark for removal list of state paths is substituted by a pattern that checks if a state is stored in a new path but the same state machine file system path and will remove duplicated states (IDs). Only the removed states will be insert in the marked for removal list or even this should be substituted by a check that clean every hierarchy level in the file system while storing.

I will give this structure a try and commit it on a separate branch.

This issue can also be used to discuss the actual situation but as long as we only have storage tests and close to no loading tests the discussion is anyway only half-grounded. 😄
The loading is one of the main reasons why I started to think about a re-factoring of this functionality.

So give me feedback if you see any big issue with this approach or your ideas to improve/simplify it.

Originally created by @Rbelder  at 2017-07-03 13:26:31+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
It is in the improve_load_and_file_system_path branch. Till now, I don't see a reason why we should not merge it.

Originally created by @Rbelder  at 2017-07-03 17:45:06+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
So I have finally removed mark for removal. Hope that is alright to everybody. As often the case we don't have test which covers this cases of removing states as I can see, but it is now quit simple and reduce complexity a lot.

Additionally I extended the limitation of the characters on the state machine path (?) and name, too. I would maybe reduce it again to only cover the state machine/library name. Therefore I would like to know your opinion.
Till now there is also a solution that covers the library names and respective changes. The changes can be batch processed by a re-save of all libraries and state machines (what should work but I have not tested it in a big scale). I insert a deprecated logger message. 

The final limitations to the names we should shortly discuss in person.

Originally created by @Rbelder  at 2017-07-04 16:53:38+00:00 (moved from RMC internal repository)
